### ISSUE TITLE
Removing vertices fails on some graphs with self loops

### ISSUE THREAD

**Issue by raoulb** [sbromberger/MetaGraphs.jl#21]
Consider the following example. I construct a graph with some boolean data on the vertices and edges. 

```
using LightGraphs
using MetaGraphs

G = Graph(6)

M = MetaGraph(G)

for (i, v) in enumerate(vertices(M))
    set_prop!(M, v, :node, isodd(i))
end

for i in 1:6
    for j in i:6
        add_edge!(M, i, j)
        set_prop!(M, i, j, :edge, isodd(i+j))
    end
end
```

Print the data:
```
for e in edges(M)
    println(e)
    println(props(M, e))
end

for v in vertices(M)
    println(v)
    println(props(M, v))
end
```

Now I try to remove some vertices

```
V = [v for v in filter_vertices(M, :node, true)]
println(V)

while length(V) > 0
    v = V[1]
    println(v)

    rem_vertex!(M, v)
    println("removed")

    V = [v for v in filter_vertices(M, :node, true)]
    println(V)
end
```

This crashes with the following out-of-bounds error:

```
ERROR: LoadError: BoundsError: attempt to access 4-element Array{Array{Int64,1},1} at index [5]                                                                                                                                                                                  
Stacktrace:                                                                                                                                                                                                                                                                      
 [1] rem_edge!(::LightGraphs.SimpleGraphs.SimpleGraph{Int64}, ::LightGraphs.SimpleGraphs.SimpleEdge{Int64}) at /home/raoul/.julia/v0.6/LightGraphs/src/graphtypes/simplegraphs/simplegraph.jl:142                                                                                
 [2] rem_vertex!(::LightGraphs.SimpleGraphs.SimpleGraph{Int64}, ::Int64) at /home/raoul/.julia/v0.6/LightGraphs/src/graphtypes/simplegraphs/SimpleGraphs.jl:105                                                                                                                  
 [3] rem_vertex!(::MetaGraphs.MetaGraph{Int64,Float64}, ::Int64) at /home/raoul/.julia/v0.6/MetaGraphs/src/MetaGraphs.jl:135                                                                                                                                                     
 [4] macro expansion at /home/raoul/LightGraphs Edge Removal Test.jl:38 [inlined]                                                                                                                                                                                                
 [5] anonymous at ./<missing>:?                                                                                                                                                                                                                                                  
 [6] include_from_node1(::String) at ./loading.jl:576                                                                                                                                                                                                                            
 [7] include(::String) at ./sysimg.jl:14                                                                                                                                                                                                                                         
 [8] process_options(::Base.JLOptions) at ./client.jl:305                                                                                                                                                                                                                        
 [9] _start() at ./client.jl:371 
```

The full output is:
```
Edge 1 => 1
Dict{Symbol,Any}(Pair{Symbol,Any}(:edge, false))
Edge 1 => 2
Dict{Symbol,Any}(Pair{Symbol,Any}(:edge, true))
Edge 1 => 3
Dict{Symbol,Any}(Pair{Symbol,Any}(:edge, false))
Edge 1 => 4
Dict{Symbol,Any}(Pair{Symbol,Any}(:edge, true))
Edge 1 => 5
Dict{Symbol,Any}(Pair{Symbol,Any}(:edge, false))
Edge 1 => 6
Dict{Symbol,Any}(Pair{Symbol,Any}(:edge, true))
Edge 2 => 2
Dict{Symbol,Any}(Pair{Symbol,Any}(:edge, false))
Edge 2 => 3
Dict{Symbol,Any}(Pair{Symbol,Any}(:edge, true))
Edge 2 => 4
Dict{Symbol,Any}(Pair{Symbol,Any}(:edge, false))
Edge 2 => 5
Dict{Symbol,Any}(Pair{Symbol,Any}(:edge, true))
Edge 2 => 6
Dict{Symbol,Any}(Pair{Symbol,Any}(:edge, false))
Edge 3 => 3
Dict{Symbol,Any}(Pair{Symbol,Any}(:edge, false))
Edge 3 => 4
Dict{Symbol,Any}(Pair{Symbol,Any}(:edge, true))
Edge 3 => 5
Dict{Symbol,Any}(Pair{Symbol,Any}(:edge, false))
Edge 3 => 6
Dict{Symbol,Any}(Pair{Symbol,Any}(:edge, true))
Edge 4 => 4
Dict{Symbol,Any}(Pair{Symbol,Any}(:edge, false))
Edge 4 => 5
Dict{Symbol,Any}(Pair{Symbol,Any}(:edge, true))
Edge 4 => 6
Dict{Symbol,Any}(Pair{Symbol,Any}(:edge, false))
Edge 5 => 5
Dict{Symbol,Any}(Pair{Symbol,Any}(:edge, false))
Edge 5 => 6
Dict{Symbol,Any}(Pair{Symbol,Any}(:edge, true))                                                                                                                                                                                                                                  
Edge 6 => 6                                                                                                                                                                                                                                                                      
Dict{Symbol,Any}(Pair{Symbol,Any}(:edge, false))                                                                                                                                                                                                                                 
1                                                                                                                                                                                                                                                                                
Dict{Symbol,Any}(Pair{Symbol,Any}(:node, true))                                                                                                                                                                                                                                  
2                                                                                                                                                                                                                                                                                
Dict{Symbol,Any}(Pair{Symbol,Any}(:node, false))                                                                                                                                                                                                                                 
3                                                                                                                                                                                                                                                                                
Dict{Symbol,Any}(Pair{Symbol,Any}(:node, true))                                                                                                                                                                                                                                  
4                                                                                                                                                                                                                                                                                
Dict{Symbol,Any}(Pair{Symbol,Any}(:node, false))                                                                                                                                                                                                                                 
5                                                                                                                                                                                                                                                                                
Dict{Symbol,Any}(Pair{Symbol,Any}(:node, true))                                                                                                                                                                                                                                  
6                                                                                                                                                                                                                                                                                
Dict{Symbol,Any}(Pair{Symbol,Any}(:node, false))                                                                                                                                                                                                                                 
[1, 3, 5]                                                                                                                                                                                                                                                                        
1                                                                                                                                                                                                                                                                                
removed                                                                                                                                                                                                                                                                          
[3, 5]                                                                                                                                                                                                                                                                           
3                                                                                                                                                                                                                                                                                
removed                                                                                                                                                                                                                                                                          
[3]                                                                                                                                                                                                                                                                              
3                                                                                                                                                                                                                                                                                
ERROR: LoadError: BoundsError: attempt to access 4-element Array{Array{Int64,1},1} at index [5]   
...        
```

I tried to be extra careful when removing vertices and not to invalidate the iterator.
Am I doing something wrong? What would be the correct way to remove all
vertices that match `filter_vertices(M, :node, true)`?

**Comment by sbromberger** [comment]
This appears to be a bug, but the best way to do this (in terms of performance, not memory) is just to induce the subgraph:
```
julia> g = PathGraph(10)
{10, 9} undirected simple Int64 graph

julia> m = MetaGraph(g)
{10, 9} undirected Int64 metagraph with Float64 weights defined by :weight (default weight 1.0)

julia> for i in vertices(m)
         set_prop!(m, i, :x, rand(Bool))
       end

julia> m[filter_vertices(m, :x, true)]
{6, 5} undirected Int64 metagraph with Float64 weights defined by :weight (default weight 1.0)
```

Will leave this open for further testing, though.

**Comment by sbromberger** [comment]
OK, the problem seems to be when you remove a vertex with self-loops. The following graph works:
```
M = MetaGraph(G)
for i in 1:5, j in i+1:6
    add_edge!(M, i, j)
    set_prop!(M, i, j, :edge, isodd(i+j))
end
```

**Comment by sbromberger** [comment]
More debugging: it looks as if not all the edges are being removed:
```
julia> V = [v for v in filter_vertices(M, :node, true)]
3-element Array{Int64,1}:
 1
 3
 5

julia> rem_vertex!(M, 1)
true

julia> V = [v for v in filter_vertices(M, :node, true)]
2-element Array{Int64,1}:
 3
 5

julia> collect(edges(M))
15-element Array{LightGraphs.SimpleGraphs.SimpleEdge{Int64},1}:
 Edge 1 => 2
 Edge 1 => 3
 Edge 1 => 4
 Edge 1 => 5
 Edge 1 => 6 # this should not be here
 Edge 2 => 2
 Edge 2 => 3
 Edge 2 => 4
 Edge 2 => 5
 Edge 3 => 3
 Edge 3 => 4
 Edge 3 => 5
 Edge 4 => 4
 Edge 4 => 5
 Edge 5 => 5

julia> V = [v for v in filter_vertices(M, :node, true)]
2-element Array{Int64,1}:
 3
 5

julia> rem_vertex!(M, 3)
true

julia> collect(edges(M))
10-element Array{LightGraphs.SimpleGraphs.SimpleEdge{Int64},1}:
 Edge 1 => 2
 Edge 1 => 3
 Edge 1 => 4
 Edge 1 => 6 # this should not be here
 Edge 2 => 2
 Edge 2 => 3
 Edge 2 => 4
 Edge 3 => 4
 Edge 3 => 5  # this should not be here
 Edge 4 => 4

julia> V = [v for v in filter_vertices(M, :node, true)]
1-element Array{Int64,1}:
 3

julia> nv(M)
4

```

**Comment by sbromberger** [comment]
This is a bug in LightGraphs, not MetaGraphs:
```
julia> g = CompleteGraph(3)
{3, 3} undirected simple Int64 graph

julia> for i in vertices(g)
         add_edge!(g, i, i)
       end

julia> collect(edges(g))
6-element Array{LightGraphs.SimpleGraphs.SimpleEdge{Int64},1}:
 Edge 1 => 1
 Edge 1 => 2
 Edge 1 => 3
 Edge 2 => 2
 Edge 2 => 3
 Edge 3 => 3

julia> rem_vertex!(g, 1)
true

julia> collect(edges(g))
3-element Array{LightGraphs.SimpleGraphs.SimpleEdge{Int64},1}:
 Edge 1 => 2
 Edge 1 => 3 # this should not be here.
 Edge 2 => 2
```

**Comment by sbromberger** [comment]
@raoulb - this should be fixed in the latest LightGraphs master. Could you confirm? Thanks.

**Comment by raoulb** [comment]
Wow, this is awesome reaction time! Thanks a lot.
Yes, the `LightGraphs` `master` seems to fix this. At least my example code does not crash anymore.

**Comment by sbromberger** [comment]
awesome. will close this out then. Thanks for reporting and testing.

**Comment by raoulb** [comment]
Hmm, while the test example works fine, there seems to be a (related?) problem on my work code.
When removing vertices from a metagraph, some edges seem to loose their metadata:

Graph data before:
```
for e in edges(Pto)
    println(e)
    println(get_prop(Pto, e, :eid))
    println(props(Pto, e))
end
```
```
Edge 1 => 6
6
Dict{Symbol,Any}(Pair{Symbol,Any}(:invisible, false),Pair{Symbol,Any}(:eid, 6),Pair{Symbol,Any}(:del, true))
Edge 1 => 13
1
Dict{Symbol,Any}(Pair{Symbol,Any}(:invisible, false),Pair{Symbol,Any}(:eid, 1),Pair{Symbol,Any}(:del, true))
Edge 1 => 14
13
Dict{Symbol,Any}(Pair{Symbol,Any}(:invisible, false),Pair{Symbol,Any}(:eid, 13),Pair{Symbol,Any}(:del, true))
Edge 2 => 8
14
Dict{Symbol,Any}(Pair{Symbol,Any}(:invisible, false),Pair{Symbol,Any}(:eid, 14),Pair{Symbol,Any}(:del, false))
Edge 2 => 13
1
Dict{Symbol,Any}(Pair{Symbol,Any}(:invisible, false),Pair{Symbol,Any}(:eid, 1),Pair{Symbol,Any}(:del, false))
Edge 2 => 15
2
Dict{Symbol,Any}(Pair{Symbol,Any}(:invisible, false),Pair{Symbol,Any}(:eid, 2),Pair{Symbol,Any}(:del, false))
Edge 3 => 4
3
Dict{Symbol,Any}(Pair{Symbol,Any}(:invisible, false),Pair{Symbol,Any}(:eid, 3),Pair{Symbol,Any}(:del, true))
Edge 3 => 15
2
Dict{Symbol,Any}(Pair{Symbol,Any}(:invisible, false),Pair{Symbol,Any}(:eid, 2),Pair{Symbol,Any}(:del, true))
Edge 3 => 16
15
Dict{Symbol,Any}(Pair{Symbol,Any}(:invisible, false),Pair{Symbol,Any}(:eid, 15),Pair{Symbol,Any}(:del, true))
Edge 4 => 5
4
Dict{Symbol,Any}(Pair{Symbol,Any}(:invisible, false),Pair{Symbol,Any}(:eid, 4),Pair{Symbol,Any}(:del, true))
Edge 4 => 10
16
Dict{Symbol,Any}(Pair{Symbol,Any}(:invisible, false),Pair{Symbol,Any}(:eid, 16),Pair{Symbol,Any}(:del, true))
Edge 5 => 6
5
Dict{Symbol,Any}(Pair{Symbol,Any}(:invisible, false),Pair{Symbol,Any}(:eid, 5),Pair{Symbol,Any}(:del, true))
Edge 5 => 11
17
Dict{Symbol,Any}(Pair{Symbol,Any}(:invisible, false),Pair{Symbol,Any}(:eid, 17),Pair{Symbol,Any}(:del, true))
Edge 6 => 12
18
Dict{Symbol,Any}(Pair{Symbol,Any}(:invisible, false),Pair{Symbol,Any}(:eid, 18),Pair{Symbol,Any}(:del, true))
Edge 7 => 8
7
Dict{Symbol,Any}(Pair{Symbol,Any}(:invisible, false),Pair{Symbol,Any}(:eid, 7),Pair{Symbol,Any}(:del, false))
Edge 7 => 14
13
Dict{Symbol,Any}(Pair{Symbol,Any}(:invisible, false),Pair{Symbol,Any}(:eid, 13),Pair{Symbol,Any}(:del, false))
Edge 7 => 17
12
Dict{Symbol,Any}(Pair{Symbol,Any}(:invisible, false),Pair{Symbol,Any}(:eid, 12),Pair{Symbol,Any}(:del, false))
Edge 8 => 9
8
Dict{Symbol,Any}(Pair{Symbol,Any}(:invisible, false),Pair{Symbol,Any}(:eid, 8),Pair{Symbol,Any}(:del, false))
Edge 9 => 16
15
Dict{Symbol,Any}(Pair{Symbol,Any}(:invisible, false),Pair{Symbol,Any}(:eid, 15),Pair{Symbol,Any}(:del, false))
Edge 9 => 18
9
Dict{Symbol,Any}(Pair{Symbol,Any}(:invisible, false),Pair{Symbol,Any}(:eid, 9),Pair{Symbol,Any}(:del, false))
Edge 10 => 11
10
Dict{Symbol,Any}(Pair{Symbol,Any}(:invisible, false),Pair{Symbol,Any}(:eid, 10),Pair{Symbol,Any}(:del, true))
Edge 10 => 18
9
Dict{Symbol,Any}(Pair{Symbol,Any}(:invisible, false),Pair{Symbol,Any}(:eid, 9),Pair{Symbol,Any}(:del, true))
Edge 11 => 12
11
Dict{Symbol,Any}(Pair{Symbol,Any}(:invisible, false),Pair{Symbol,Any}(:eid, 11),Pair{Symbol,Any}(:del, true))
Edge 12 => 17
12
Dict{Symbol,Any}(Pair{Symbol,Any}(:invisible, false),Pair{Symbol,Any}(:eid, 12),Pair{Symbol,Any}(:del, true))
```

Then, I try to delete some parts (still using this dubious method):

```
function finish_topology(Pt)
    V = [v for v in filter_vertices(Pt, :del, true)]
    while length(V) > 0
        v = V[end]
        println(v)
        rem_vertex!(Pt, v)
        println("removed")
        V = [v for v in filter_vertices(Pt, :del, true)]
        println(V)
    end
end
finish_topology(Pto)
```
```
12
removed
[1, 3, 4, 5, 6, 10, 11]
11
removed
[1, 3, 4, 5, 6, 10]
10
removed
[1, 3, 4, 5, 6]
6
removed
[1, 3, 4, 5]
5
removed
[1, 3, 4]
4
removed
[1, 3]
3
removed
[1]
1
removed
Int64[]
```

And what I end up with is:

```
for e in edges(Pto)
    println(e)
    #println(get_prop(Pto, e, :eid))
    println(props(Pto, e))
end
```
```
Edge 1 => 7
Dict{Symbol,Any}()
Edge 2 => 4
Dict{Symbol,Any}()
Edge 2 => 6
Dict{Symbol,Any}()
Edge 2 => 8
Dict{Symbol,Any}(Pair{Symbol,Any}(:invisible, false),Pair{Symbol,Any}(:eid, 14),Pair{Symbol,Any}(:del, false))
Edge 3 => 9
Dict{Symbol,Any}()
Edge 5 => 7
Dict{Symbol,Any}()
Edge 7 => 8
Dict{Symbol,Any}(Pair{Symbol,Any}(:invisible, false),Pair{Symbol,Any}(:eid, 7),Pair{Symbol,Any}(:del, false))
Edge 8 => 9
Dict{Symbol,Any}(Pair{Symbol,Any}(:invisible, false),Pair{Symbol,Any}(:eid, 8),Pair{Symbol,Any}(:del, false))
Edge 9 => 10
Dict{Symbol,Any}()
```

Clearly, some data got lost somewhere.

**Comment by sbromberger** [comment]
Can you supply a MWE for me to test?

**Comment by raoulb** [comment]
Using this implementation based on your proposal:
```
function finish_topology(Pt)
    return Pt[filter_vertices(Pt, :del, false)]
end
```
works fine however.

**Comment by raoulb** [comment]
Hmm, the MWE is what I tried within the first report. I think I could upload a jupyter notebook.
The code is highly experimental and messy however.

**Comment by sbromberger** [comment]
I need to be able to reproduce the bug.

**Comment by sbromberger** [comment]
Ah, wait. you're missing edge properties, right?

**Comment by sbromberger** [comment]
I don't see a good way to fix this, unfortunately. Keep using the induced subgraph technique while I think about it a bit.

cc @jpfairbanks - the issue is that `rem_vertex!` with MetaGraphs calls `rem_edge!` and `add_edge!` for the underlying simplegraph but neither of those preserve the metadata. We'll have to override `rem_vertex!` somehow to do this.

**Comment by raoulb** [comment]
This should be minimal:
https://gist.github.com/raoulb/a2013cfb2449e5ad4babed7d9f16ccf1

**Comment by sbromberger** [comment]
That's hardly minimal, but I think I've identified the problem.

**Comment by raoulb** [comment]
I know, but as minimal as I could get it based on my jupyter notebook.

**Comment by sbromberger** [comment]
I opened #22 to track this and I fixed it via #29. Closing out. If you still see problems here, please let me know!
